ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/core/tests/internal/PDEXmlProcessorFactoryTest.java

```diff
@@ -0,0 +1,344 @@
+			Document document;
+
+			try (InputStream xmlStream = xmlSupplier.apply(httpServerThread.getLocalPort())) {
+				document = builder.parse(xmlStream);
+			}
+			Element root = document.getDocumentElement();
+
+			assertEquals("Body", root.getTagName());
+			String value = root.getChildNodes().item(0).getNodeValue();
+			assertFalse("Parser injected secret: " + value, value.contains("secret"));
+		} finally {
+			cleanup();
+		}
+	}
+
+	@Test
+	public void testTransformXmlWithExternalEntity() throws Exception {
+		TransformerFactory transformerFactory = PDEXmlProcessorFactory.createTransformerFactoryWithErrorOnDOCTYPE();
+		try {
+			testParseXmlWithExternalEntity(transformerFactory, this::createMalciousXml);
+			assertTrue("TransformerException expected", false);
+		} catch (TransformerException e) {
+			String message = e.getMessage();
+			assertTrue(message, message.contains("DTD"));
+		}
+	}
+
+	@Test
+	public void testTransformXmlWithoutExternalEntity() throws Exception {
+		TransformerFactory transformerFactory = PDEXmlProcessorFactory.createTransformerFactoryWithErrorOnDOCTYPE();
+		testParseXmlWithExternalEntity(transformerFactory, this::createNormalXml);
+	}
+
+	public void testParseXmlWithExternalEntity(TransformerFactory transformerFactory,
+			IntFunction<InputStream> xmlSupplier) throws Exception {
+		try (Server httpServerThread = new Server()) {
+			String formatted;
+
+			try (InputStream xmlStream = xmlSupplier.apply(httpServerThread.getLocalPort())) {
+				Transformer xformer = transformerFactory.newTransformer();
+				Source source = new StreamSource(xmlStream);
+				try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+					Result result = new StreamResult(outputStream);
+					xformer.transform(source, result);
+					formatted = outputStream.toString(StandardCharsets.UTF_8);
+				}
+			}
+			assertTrue(formatted, formatted.contains("<Body>"));
+			assertFalse("Formatter injected secret: " + formatted, formatted.contains("secret"));
+		} finally {
+			cleanup();
+		}
+	}
+
+	private InputStream createMalciousXml(int localPort) {
+		try {
+			Path tempSecret = Files.createTempFile("test", ".txt");
+			tmpFiles.add(tempSecret);
+			Files.writeString(tempSecret, "secret");
+			Path tempDtd = Files.createTempFile("test", ".dtd");
+			tmpFiles.add(tempDtd);
+			String dtdContent = "<!ENTITY % var1 SYSTEM \"" + tempSecret.toUri().toURL() + "\">\n" //
+					+ "<!ENTITY var4 SYSTEM \"" + tempSecret.toUri().toURL() + "\">\n" //
+					+ "<!ENTITY % var2 \"<!ENTITY var3 SYSTEM 'http://localhost:" + localPort + "/?%var1;'>\">\n" //
+					+ "%var2;\n";
+			Files.writeString(tempDtd, dtdContent);
+			StringBuilder sb = new StringBuilder();
+			sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
+			URI dtdURi = tempDtd.toUri();
+			String dtdUrl = dtdURi.toURL().toString();
+			sb.append("<!DOCTYPE var1 SYSTEM \"" + dtdUrl + "\">\n");
+			sb.append("<Body>&var3;&var4;</Body>");
+			String xml = sb.toString();
```

Comment:
Text-Block + String.format again:
```suggestion
			String xml = String.format("""
					<?xml version="1.0" encoding="UTF-8"?>
					<!DOCTYPE var1 SYSTEM "%s">
					<Body>&var3;&var4;</Body>""", tempDtd.toUri().toURL());
```

Reply:
i am not familiar to read String.format

Reply:
OK. But then at least the same style (with regards to StringBuilder vs. String concatenation) should be used.
In my extra commit I changed it to use String concatenation.